releasenotes/notes/add-elide-swaps-b0a4c373c9af1efd.yaml

```diff
@@ -0,0 +1,47 @@
+---
+features:
+  - |
+    Added a new optimization transpiler pass, :class:`~.ElidePermutations`,
+    which is designed to run prior to the :ref:`layout_stage` and will
+    optimize away any :class:`~.SwapGate`\s and
+    :class:`~qiskit.circuit.library.PermutationGate`\s
+    in a circuit by permuting virtual
+    qubits. For example, taking a circuit with :class:`~.SwapGate`\s:
+
+    .. plot::
+
+       from qiskit.circuit import QuantumCircuit
+
+       qc = QuantumCircuit(3)
+       qc.h(0)
+       qc.swap(0, 1)
+       qc.swap(2, 0)
+       qc.cx(1, 0)
+       qc.measure_all()
+       qc.draw("mpl")
+
+    will remove the swaps when the pass is run:
+
+    .. plot::
+       :include-source:
+
+       from qiskit.transpiler.passes import ElidePermutations
+       from qiskit.circuit import QuantumCircuit
+
+       qc = QuantumCircuit(3)
+       qc.h(0)
+       qc.swap(0, 1)
+       qc.swap(2, 0)
+       qc.cx(1, 0)
+       qc.measure_all()
+
+       ElidePermutations()(qc).draw("mpl")
+
+    The pass also sets the ``virtual_permutation_layout`` property set, storing
+    the permutation of the virtual qubits that was optimized away.
+
+  - |
+    Added a new transpiler pass, :class:`~.FinalizeLayouts`,
+    which is designed to run after the :ref:`routing_stage` and update the
+    ``final_layout`` property set based on ``layout`` and
+    ``virtual_permutation_layout``.
```

Comment:
From your discussion above @jakelishman , it sounds to me like we ought to consider keeping any new properties private until we i.e. refactor to storing `final_permutation`.

Thoughts?

Reply:
I'm fine not documenting it explicitly as part of the API although at the end of the day the pass doesn't do much in the absence of a virtual permutation layout. So it's kind of hard to keep it private as part of the pass. I don't think it's adding an extra burden realistically to support it for the remainder of 1.x if we pivot to something else in the future we have to do this already with `final_layout`.

Reply:
Yeah, part of us not being able to get the `final_permutation` landed properly for 1.1 is that basically have to support this for the whole 1.x cycle now, but that was priced into the decision - it shouldn't be _too_ painful, and we'd have needed a 2.0 release to make a truly clean break from external routing passes anyway.